test/e2e/network/service.go

```diff
@@ -4226,7 +4227,81 @@ var _ = common.SIGDescribe("Services", func() {
 		service, err := t.CreateService(service)
 		framework.ExpectNoError(err)
 
-		checkServiceReachabilityFromExecPod(ctx, f.ClientSet, ns, service.Name, service.Spec.ClusterIP, port)
+		_ = checkServiceReachabilityFromExecPod(ctx, f.ClientSet, ns, service.Name, service.Spec.ClusterIP, port)
+	})
+
+	ginkgo.It("should connect to the named ports during port number changing", func(ctx context.Context) {
+		serviceName := "mutable-named-port"
+		ns := f.Namespace.Name
+
+		t := NewServerTest(cs, ns, serviceName)
+		defer func() {
+			defer ginkgo.GinkgoRecover()
+			errs := t.Cleanup()
+			if len(errs) != 0 {
+				framework.Failf("errors in cleanup: %v", errs)
+			}
+		}()
+
+		port := int32(8080)
+		portName := "mutable-port"
+		ports := []v1.ContainerPort{{Name: portName, ContainerPort: port, Protocol: v1.ProtocolTCP}}
+		args := []string{"netexec", fmt.Sprintf("--http-port=%d", port)}
+		pods := []*v1.Pod{}
+		for _, podName := range []string{"testpod0", "testpod1"} {
+			pods = append(pods, createPodOrFail(ctx, f, ns, podName, t.Labels, ports, args...))
+		}
+
+		service := t.BuildServiceSpec()
+		servicePort := int32(8001)
+		service.Spec.Ports = []v1.ServicePort{
+			{
+				Name:       portName,
+				Port:       servicePort,
+				TargetPort: intstr.FromString(portName),
+			},
+		}
+
+		ginkgo.By(fmt.Sprintf("creating Service %v with selectors %v", service.Name, service.Spec.Selector))
+		service, err := t.CreateService(service)
+		framework.ExpectNoError(err)
+
+		getPodNamesBehindService := func(expectPodCnt int) (podNames []string) {
+			for retryCnt := 0; retryCnt < max(20, expectPodCnt*2); retryCnt++ {
+				output := checkServiceReachabilityFromExecPod(ctx, f.ClientSet, ns, service.Name, service.Spec.ClusterIP, servicePort)
+				if !slices.Contains(podNames, output) {
+					podNames = append(podNames, output)
+				}
+			}
+			slices.Sort(podNames)
+			return podNames
+		}
```

Comment:
I'd mentioned in the issue that you can reuse the session affinity helpers for this. In particular, if you call:

    reachedAllEndpoints := checkAffinity(ctx, f.ClientSet, execPod, service.Spec.ClusterIP, servicePort, false)

then that will repeatedly connect to the service (with appropriate timeouts and such), until it has successfully reached more than one endpoint. (The `false` at the end means "make sure affinity doesn't hold".) So then you don't need to poll and compare hostnames and all that yourself.

(To make `checkAffinity` work you'll need to be running `serve-hostname` rather than `netexec` in the pods. And you'll have to create your own `execPod`, but you can copy the code for that from elsewhere.)